File: code/python/src/rl_agent/run2.py
"""
Alternative execution script for reinforcement learning trading strategy.

This script provides an alternative implementation with additional visualization
and data handling capabilities.
"""

import matplotlib.pyplot as plt
import pandas as pd
from src.rl_agent_python.attributes import attributes_all
from src.rl_agent_python.cointegration import get_pairs
from src.rl_agent_python.financial_functions import calculate_returns, plot_agent_result
from src.rl_agent_python.mcc import mcc_bootstrap
from src.rl_agent_python.pre_run import pre_run
from src.rl_agent_python.q_learning import q_control


def plot_time_series(data: pd.DataFrame, title: str = "Time Series") -> None:
    """
    Plot time series data.

    Args:
        data: DataFrame containing time series data
        title: Plot title
    """
    plt.figure(figsize=(12, 6))
    plt.plot(data)
    plt.title(title)
    plt.xlabel("Time")
    plt.ylabel("Value")
    plt.grid(True)
    plt.show()


def main():
    # Initialize data
    currencies = [
        "EURGBP",
        "EURUSD",
        "USDJPY",
        "EURJPY",
        "CHFJPY",
        "GBPUSD",
        "AUDJPY",
        "EURCAD",
    ]

    # Get currency data
    df_currencies = pre_run(currencies)

    # Define training and testing periods
    train_period = (1, 1000)
    test_period = (1001, 1300)  # 300 days of testing

    # Get pairs data
    pairs = get_pairs(df_currencies, currencies, train_period)
    train_pairs = pd.concat(pairs[0], axis=1)
    test_pairs = pd.concat(pairs[1], axis=1)
    all_pairs = pd.concat(pairs[2], axis=1)

    # Create attributes for all pairs
    delay = 2
    attributes_dict = attributes_all(all_pairs, all_pairs.columns, delay)

    # Select a pair for testing
    pair_idx = 0

    # Prepare training data
    train_pair = train_pairs.iloc[:, pair_idx]
    train_features = pd.DataFrame(attributes_dict[pair_idx]).iloc[
        train_period[0] - 1 : train_period[1]
    ]

    # Prepare testing data
    test_pair = test_pairs.iloc[:, pair_idx]
    test_features = pd.DataFrame(attributes_dict[pair_idx]).iloc[
        test_period[0] - 1 : test_period[1]
    ]

    # Visualize data
    plot_time_series(train_pair, "Training Data")
    plot_time_series(test_pair, "Testing Data")

    # Test Monte Carlo Control
    n_episodes = 10
    cost = 0

    # Train MCC agent
    pretrained_agent = mcc_bootstrap(
        train_features,
        train_pair,
        None,  # coef will be calculated internally
        cost,
        n_episodes,
        algorithm="mcc",
    )

    pretrained_pos = pretrained_agent.Position_track
    pretrain_return = calculate_returns(
        train_pair, None, pretrained_pos[-len(train_pair) :], cost
    )
    plot_agent_result(train_pair, None, pretrain_return, "Monte Carlo Control")

    # Test MCC agent
    tested_agent = mcc_bootstrap(
        test_features,
        test_pair,
        None,  # coef will be calculated internally
        cost,
        n_episodes,
        pretrained_agent=pretrained_agent,
    )

    test_pos = tested_agent.Position_track
    test_return = calculate_returns(test_pair, None, test_pos[-len(test_pair) :], cost)
    plot_agent_result(test_pair, None, test_return, "Monte Carlo Control")

    # Test Q-Learning
    n_episodes = 5
    cost = 0

    # Train Q-Learning agent
    pretrained_agent = q_control(
        train_pair,
        None,  # coef will be calculated internally
        train_features,
        n_episodes,
        cost,
        verbose=True,
        reward_function="dsr",
    )

    pretrained_pos = pretrained_agent[1]  # Position track is second element
    pretrain_return = calculate_returns(
        train_pair, None, pretrained_pos[-len(train_pair) :], cost
    )
    plot_agent_result(train_pair, None, pretrain_return, "Q-Learning Control")

    # Test Q-Learning agent
    tested_agent = q_control(
        test_pair,
        None,  # coef will be calculated internally
        test_features,
        1,  # Single episode for testing
        cost,
        pretrained_agent=pretrained_agent,
    )

    test_pos = tested_agent[1]
    test_return = calculate_returns(test_pair, None, test_pos[-len(test_pair) :], cost)
    plot_agent_result(test_pair, None, test_return, "Q-Learning Control")


if __name__ == "__main__":
    main()

File: code/R/src/rl_agent/run2.R
source('./src/rl_agent/pre_run.R')

#c_currencies <- list.files("C:/Users/user/Desktop/UCL final project/data/1min") # 1min data
c_currencies <- c("EURGBP", "EURUSD", "USDJPY", "EURJPY", "CHFJPY", "GBPUSD", "AUDJPY", "EURCAD")
#c_currencies <- randomObs(c_currencies, 12) # 12 randon currencies
#c_currencies <- c_currencies[! c_currencies %in% c(c_currencies1)]
#df_currencies <- GetQuotes(c_currencies, data_path, ".csv")

# New data 08.2015
#c_currencies <- c("EURGBPask", "EURUSDask", "USDJPYask", "EURJPYask", "CHFJPYask", "GBPUSDask", "AUDJPYask", "EURCADask")
df_currencies <- GetQuotes(c_currencies, data_path, ".csv")

#endregion
# -----------------------------------------------------------------
#region Cointegration extraction

# Get all cointegrated pairs on a specified base interval

train_period <- c(1,1000)
test_period <- c(1001,300)
debugonce(GetCointPairs)

pairs <- GetPairs(df_currencies, c_currencies, train_period) %>%
  purrr::transpose() %>%
  lapply(function(x) invoke(cbind, x) %>% as.data.frame())



#endregion
# -----------------------------------------------------------------
#region Attributes extraction

#Create the data frame of attributes for all cointegrated pairs
i_delay <- 2
l_attributes <- AttributesAll(pairs[[3]], colnames(pairs[[3]]), i_delay)
str(l_attributes)
#endregion
# -----------------------------------------------------------------
#region RL backtest data preparation

# Pick one of the pairs and it's attributes

pair_ind <- 1

length(l_attributes[[3]])
train_pair <- pairs[[1]][, pair_ind]
train_features <- as.data.frame(l_attributes[pair_ind])[train_period[1]:train_period[2], ]

test_pair <- pairs[[2]][, pair_ind]
test_features <- as.data.frame(l_attributes[pair_ind])[test_period[1]:test_period[2],]

plot.ts(test_pair)
plot.ts(train_pair)

#endregion
# -----------------------------------------------------------------
#region MCC test





Nepisodes <- 10
cost <- 0
debugonce(MCCbootstrap)
pretrained_agent <- MCCbootstrap(coint)


pretrained_pos <- pretrained_agent[[2]]
pretrain_return <- Ret(train_pair, coef, tail(pretrained_pos, nrow(train_pair)), 0)
PlotAgentResult(train_pair, train_spread, pretrain_return,'Monte Carlo Control')


tested_agent <- MCCbootstrap(test_spread, test_pair, coef, test_features, Nepisodes, cost, pretrained_agent)
test_pos <- tested_agent[[2]]
test_return <- Ret(test_pair, coef, tail(test_pos, nrow(test_pair)), 0)
PlotAgentResult(test_pair, test_spread, test_return, 'Monte Carlo Control')

# PRZEKAZANIE VSA DO TESTOWEGO OKRESU. Jak to ma jakies parametry typu alpha to to powinniem byc hyper parametr
# jezeli mamy zamiar ogolnie optymalizowac. Czy dlugosc epizodu i mierzenia kointegracji (base i test period) mogla
# by byc wstawiona jako parametry dla agenta?

#endregion
# -----------------------------------------------------------------
#region Qlearning test

Nepisodes <- 5
cost <- 0

#experience replay - nie przejmuje sie co sie dzieje do moment uduzego skoku, a potem juz tylko idzie w gore.

pretrained_agent <- Qcontrol(train_pair, coef, train_features, Nepisodes, cost)
pretrained_pos <- pretrained_agent[[2]]
pretrain_return <- Ret(train_pair, coef, tail(pretrained_pos, nrow(train_pair)), 0)
PlotAgentResult(train_pair, train_spread, pretrain_return,'Qlearning Control')


tested_agent <- Qcontrol(test_pair, coef, test_features, 1, cost, pretrained_agent)
test_pos <- tested_agent[[2]]
test_return <- Ret(test_pair, coef, tail(test_pos, nrow(test_pair)), 0)
PlotAgentResult(test_pair, test_spread, test_return, 'Qlearning Control')
